content/blog/2019-10-21-threads-not-as-libraries/index.md

```diff
@@ -0,0 +1,272 @@
+It is in this sense that **Threads Cannot Be Implemented as a Library**: if the language doesn't specify the behavior of threads, optimizations cannot be guaranteed to preserve the thread library's specification, even if they sometimes do in practice.
+
+
+
+## Three Optimizations that Break Pthreads
+The authors identify three compiler optimizations that break otherwise well-defined Pthreads programs:
+
+### Concurrent modification
+
+Consider the following two statements which are run on 2 separate threads with both the variables initialized to zero:
+
+```c
+// Thread 1
+if (x==1) ++y;
+
+// Thread 2
+if (y==1) ++x;
+```
+
+Since there is no way under SC that either `x` or `y` can be read and modified concurrently, this program does not contain data races, and is therefore well-defined.
+
+Since the compiler can perform any optimization that preserves 'single-threaded' correctness, it may opt to perform a "speculative" optimization as follows:
+```c
+// Thread 1
+++y; if (x != 1) --y;
+
+// Thread 2
+++x; if (y != 1) --x;
+```
+
+This introduces a data race, since both `x` and `y` are concurrently read and modified. 
+Thus, the originally well-defined program is transformed into one that is undefined.
+
+### Rewriting of Adjacent Data
+
+Given the following struct definition:
+
+```c
+struct {int a:17; int b:15} x;
+```
+
+A compiler may generate a data race from the following concurrent field assignments:
+
+```c
+x.a = 42; //thread1
+...
+x.b = 37; //thread2
+```
+
+At first glance, one might expect the above program to be race free, since both concurrent writes appear to access distinct variables.
+However, Pthreads does not define data races in terms of *program variables*, but rather in terms of **memory locations**.
+Since this term is not formally defined by Pthreads, compilers may have different program variables unexpectedly *share* a memory location.
+
+In the case above, a compiler targeting a 32-bit machine may choose to have the fields `x.a` and `x.b` share a memory location. This is probable in fact, since such machines are likely to have a 32-bit wide, rather than a 17 or 15-bit wide, store operation, necessitating the following implementation of bit-field assignment for `x.a = 42`:
+
+```c
+{
+  tmp = x;
+  tmp &= ~0x0001ffff;
+  tmp |= 42;
+  x = tmp;
+}
+```
+
+Thus, both fields `x.a` and `x.b` share memory location `x`, causing a data race when concurrently modified.
+
+As mentioned, this problematic transformation over adjacent bit-fields is motivated by architectural constraints. 
+However, a compiler can also deploy this transformation as a memory-saving optimization over adjacent fields.
+
+For example, given the following structure definition:
+```c
+struct { char a; char b; char c; char d; char e; char f; char g; char h; } x;
+```
+
+And a program that concurrently writes to adjacent fields:
+```c
+// Thread 1
+x.a = ‘a’ 
+
+ // Thread 2
+x.b = ‘b’; x.c=’c’; x.d=’d’; x.e=’e’; x.f=’f’; x.g=’g’; x.h=’h’;
```

Comment:
Be sure to use "straight" quotes in code. `x.b = ‘b’;` won't parse in any C compiler because it uses "curly" quotes.